sklearn/ensemble/_hist_gradient_boosting/_loss.pyx

```diff
@@ -49,44 +91,66 @@ def _update_gradients_hessians_binary_crossentropy(
-        p_i = _cexpit(raw_predictions[i])
-        gradients[i] = p_i - y_true[i]
-        hessians[i] = p_i * (1. - p_i)
+    if sample_weight is None:
+        for i in prange(n_samples, schedule='static', nogil=True):
+            p_i = _cexpit(raw_predictions[i])
+            gradients[i] = p_i - y_true[i]
+            hessians[i] = p_i * (1. - p_i)
+    else:
+        for i in prange(n_samples, schedule='static', nogil=True):
+            p_i = _cexpit(raw_predictions[i])
+            gradients[i] = (p_i - y_true[i]) * sample_weight[i]
+            hessians[i] = p_i * (1. - p_i) * sample_weight[i]
 
 
 def _update_gradients_hessians_categorical_crossentropy(
         G_H_DTYPE_C [:, ::1] gradients,  # OUT
         G_H_DTYPE_C [:, ::1] hessians,  # OUT
         const Y_DTYPE_C [::1] y_true,  # IN
-        const Y_DTYPE_C [:, ::1] raw_predictions):  # IN
+        const Y_DTYPE_C [:, ::1] raw_predictions,  # IN
+        const Y_DTYPE_C [::1] sample_weight):  # IN
     cdef:
         int prediction_dim = raw_predictions.shape[0]
         int n_samples = raw_predictions.shape[1]
         int k  # class index
         int i  # sample index
+        Y_DTYPE_C sw
         # p[i, k] is the probability that class(ith sample) == k.
         # It's the softmax of the raw predictions
         Y_DTYPE_C [:, ::1] p = np.empty(shape=(n_samples, prediction_dim))
         Y_DTYPE_C p_i_k
 
-    for i in prange(n_samples, schedule='static', nogil=True):
-        # first compute softmaxes of sample i for each class
-        for k in range(prediction_dim):
-            p[i, k] = raw_predictions[k, i]  # prepare softmax
-        _compute_softmax(p, i)
-        # then update gradients and hessians
-        for k in range(prediction_dim):
-            p_i_k = p[i, k]
-            gradients[k, i] = p_i_k - (y_true[i] == k)
-            hessians[k, i] = p_i_k * (1. - p_i_k)
+    if sample_weight is None:
+        for i in prange(n_samples, schedule='static', nogil=True):
+            # first compute softmaxes of sample i for each class
+            for k in range(prediction_dim):
+                p[i, k] = raw_predictions[k, i]  # prepare softmax
+            _compute_softmax(p, i)
+            # then update gradients and hessians
+            for k in range(prediction_dim):
+                p_i_k = p[i, k]
+                gradients[k, i] = p_i_k - (y_true[i] == k)
+                hessians[k, i] = p_i_k * (1. - p_i_k)
+    else:
+        for i in prange(n_samples, schedule='static', nogil=True):
+            # first compute softmaxes of sample i for each class
+            for k in range(prediction_dim):
+                p[i, k] = raw_predictions[k, i]  # prepare softmax
+            _compute_softmax(p, i)
+            # then update gradients and hessians
+            sw = sample_weight[i]
+            for k in range(prediction_dim):
+                p_i_k = p[i, k]
+                gradients[k, i] = (p_i_k - (y_true[i] == k)) * sw
```

Comment:
There does not seem to any python interactions with:
```suggestion
                gradients[k, i] = (p_i_k - (y_true[i] == k)) * sample_weight[i]
```

Reply:
I think it's more about making 1 array access vs `2 * k` accesses

Reply:
Ah okay